man/man1/find-filter.1
'\"macro stdmacro
.TH FIND-FILTER 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3find-filter\f1 \- fine-grained file selection based on time attributes
.SH SYNOPSIS
\f3$PCP_BINADM_DIR/find-filter\f1
[\f3\-v?\f1]
\f2predicate\f1
.SH DESCRIPTION
.B find-filter
is a filter for a list
of file names read on standard input, and if the files match the
.I predicate
their names are written on standard output.
.PP
The supported
.I predicate
forms are based on the file's creation
time or modification time, and take the form
.B ctime
or
.B mtime
followed by a time specification.
A time specification takes the form of a \f3+\f1 or
\f3-\f1 followed by days (a non-negative integer),
optionally followed by a colon (\f3:\f1) and hours
(an integer in the range 0 to 23),
optionally followed by another colon (\f3:\f1)
and minutes (an integer in the range 0 to 59).
.PP
The semantics of the time specification are that a file matches
the predicate if the
chosen time attribute for the file is
.I "less than"
(\f3+\f1) the current time
minus the time specification, else
.I "more than or equal to"
(\f3-\f1) the current time
minus the time specification.
Alternatively, \f3+\f1 means
.I before
the current time minus the
time specification, and
\f3-\f1 means
.I at
or
.I after
the current time minus the
time specification.
.PP
.B find-filter
is intended to be used to provide finer-grain and platform independent
selection compared to the
.B \-mtime
or
.B \-ctime
options of
.BR find (1).
.B find-filter
is expected to be used as a post-filter for the output from
.BR find (1),
and this is how it is used in the
.BR pmlogger_daily (1)
scripts to select files to be compressed or culled.
.SH EXAMPLES
The following
.I predicate
examples illustrate some typical uses.
.IP "\fBmtime +3" 12n
Matches files modified more than 3 days ago.
.IP "\fBctime -0:2" 12n
Matches files created in the last 2 hours.
.IP "\fBctime +0:0:15" 12n
Matches files created more than 15 minutes ago.
.SH DIAGNOSTICS
There are none, unless the
.B \-v
option is used, in which case some diagnostics are written to standard
error.
.PP
Exit status is always zero, unless there are errors found
when parsing the command line arguments.
.SH SEE ALSO
.BR find (1)
and
.BR pmlogger_daily (1).

.\" control lines for scripts/man-spell
.\" +ok+ mtime


CHANGELOG
pcp-6.2.2 (29 May 2024)
    - Work-in-progress, not yet released.

pcp-6.2.1 (10 April 2024)
    - PMDA additions, enhancements and bug fixes:
	pmdauwsgi: (new) instrumentation from uWSGI servers
	pmdalinux: new hugepage metrics from sysfs (fixed size hugepages)
	pmdalinux: new filesys.uuid and filesys.type metrics
	pmdalinux: new TcpExt metrics from /proc/net/netstat
	pmdalinux: new softnet metrics from newer kernel versions
	pmdalinux: new /proc/vmstat khugepaged metrics
	pmdabpf: updates to add eBPF networking metrics for pcp-atop
	pmdaopenmetrics: fix script error diagnostic cascading exception
    - Client tools and utilities:
	pcp2openmetrics: (new) push PCP metrics in OpenMetrics format
	pmcheck: (new) interrogate and control PCP components
	pmcheck: Redis server detection and agent recommendation
	pmcheck: PostgreSQL server detection and agent recommendation
	pmcheck: uwSGI server detection and agent recommendation
	pcp-atopsar: fix TZ mishandling relating to day-crossover
	pcp-dstat: fix an I/O config typo, sum reads and writes
	pcp-htop: sync latest htop-3.3.0 fixes
	pmrep.conf: use bpf instead of bcc pmda for proc net metrics
	pmrep.conf: proc-essential cosmetic improvements
	pmrep.conf: remove vmeff field from sar -B config
	pmrepconf: use pmrep-internal representations for each metric
    - Server-side utilities and log management scripts:
	pmlogger: bizarre timezone fix (github #1936)
	pmlogger: improve handling of a change in metadata semantics
	pmlogdump: add support for multi-archive contexts
	pmproxy: disable Redis protocol proxying by default
	runaspcp: remove hard-coded "pcp" for user and group
	init scripts: drop conditonal use of setpriv(1) or runuser(1)
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix logmeta.c SEGV with V3 archives
	libpcp: fix memleak on repeated __pmFixPMNSHashTab calls
	libpcp: add pmgetopt pmflush() call for warnings
	libpcp: fixes for V3 archives and multi-archive contexts
	libpcp: export __pmLogChangeArchive() interface to tools
	libpcp: fix for PMNS with multi-archive contexts
	libpcp: small derived metrics bug fix (github #1921)
    - Build, infrastructure and packaging updates:
	Linux MX distro support (Debian without systemd)
	riscv: configure pmdabpf_arch value correctly
	tar packages: fix perl modules install in postinstall script
	eBPF updates: add submodule for blazesym, sync to latest
	debian build: add postrm for pcp-zeroconf
	build: removed cppcheck from 'make check', too flakey
	container: Fedora 40 container testing added
	constainer: fixed Fedora base image location (quay.io)
	build: add support for riscv64 to pcp rpm spec files
    - Security Enhanced Linux:
	selinux policy: small tweak for pmie signal handling
    - Documentation:
	man pages: document pmlogger_check and pmlogger_daily SaveLogs
	man pages: man-spell and corrections
	docs: updated donations page with new fiscal sponsor details

pcp-6.2.0 (12 February 2024)
    - Client tools and utilities:
	pmlogger: default to creating version 3 PCP archives
	pmie, pmlogger, pmcd, pmproxy: init script hardening to
	    run less as root:root, more as pcp:pcp and more safely
	pcp-reboot-init: new systemd service for $PCP_RC_DIR setup
	pcp2arrow: new, export metrics to parquet format for pandas
	pcp-atop: deterministic process sort when handling NULLs
	pcp-atop: update to v2.10.0 upstream code
	pcp-htop: update to latest upstream fixes
	pcp-geolocate: function with older versions of python
	pcp-meminfo: resolve possible 'broken pipe' diagnostics
	pcp-ss: fix handling of listening state sockets to match ss
	pcp-ss: fix the UDP UNCONN vs LISTEN state handling
    - PMDA additions, enhancements and bug fixes:
	pmdaproc: improvements for process accounting metrics
	pmdalinux: add pgpromote* and pgdemote* memory metrics
	pmdalinux: add metric sysfs.module.zswap.max_pool_percent
	pmdalinux: add new SNMP metric network.ip.outtransmits
	pmdasummary: send state change 0 (no change) back to pmcd
	pmdapostgresql: support the psycopg3 python module as well
	pmdaopenmetrics: kepler support - live metrics and logging
    - Server-side utilities and log management scripts:
	pmlogrewrite, pmlogredact: fix for missing data volumes
	pmcd: cosmetic change to avoid pipe2 name clash
	runaspcp: new helper utility for unprivileged rc scripts
	pmnsmerge: be more defensive in qsort comparison routine
	pmlogsize: fix qsort comparison function handling of equality
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: pmOpenLog refactor for handling previous log files
	libpcp: new __pmCleanMapDir routine to aid log culling
	libpcp & pmcd: READY-NOT-READY hardening
	libpcp: avoid possible thread issues (coverity)
	libpcp: refactor pmDupContext for thread safety
	libpcp_import: add guard in pmiPutMark to prevent SIGSEGV
    - Build, infrastructure and packaging updates:
	build: changes to re-enable builds for FreeBSD 12, 13, 14
	build: add diffutils rpm dep as pmdaproc script uses diff
	build: add main.yml for building and publishing container
	build: resolve compiler warnings in various tools/libraries
	packaging: use tmpfiles.d method to create PCP_RUN_DIR
	build: packaging changes for openSUSE 15.5 and 15.6
	debian: do not ship empty directory /usr/lib/pkgconfig
	build: el10 systemd-presets for pmcd/pmie/pmlogger enabling
	build: drop rpm dep on bpftool as local version used only
	build: change perl YAML::XS::LibYAML module to YAML::XS
    - Security Enhanced Linux:
	selinux: policy improvements for CentOS/RHEL 7 and 8
    - Documentation:
	docs: Fedora 39 found some tbl(1)/troff(1) issues, fix
	docs: add an explicit code-of-conduct for PCP
	docs: pmns(5) note that metric names must only contain ASCII
	docs: update the BPF README to describe bpftool reality

pcp-6.1.1 (17 November 2023)
    - Client tools and utilities:
	pmie: add -o/--format to control output format with archives

src/find-filter/find-filter.c
/*
 * Copyright (c) 2020 Ken McDonell.  All Rights Reserved.
 */
#include "pmapi.h"
#include <sys/stat.h>

static void
usage(void) {
    fprintf(stderr, "Usage: find-filter [options] <predicate>\n");
    fprintf(stderr, "options:\n");
    fprintf(stderr, "  -v                 verbose\n");
    fprintf(stderr, "\n");
    fprintf(stderr, "<predicate> is one of the following:\n");
    fprintf(stderr, "  ctime <timespec>   filter on a file's ctime\n");
    fprintf(stderr, "  mtime <timespec>   filter on a file's mtime\n");
    fprintf(stderr, "and <timespec> is one of the following:\n");
    fprintf(stderr, "  +D[:H[:M]]         file's time is > specified days, hours and minutes ago\n");
    fprintf(stderr, "  -D[:H[:M]]         file's time is <= specified days, hours and minutes ago\n");
    exit(1);
}

#define PRED_NONE	-1
#define PRED_CTIME	1
#define PRED_MTIME	2

#define PLUS_TIME	1
#define MINUS_TIME	2

int
main(int argc, char **argv)
{
    struct timeval	now;
    double		now_f;
    struct timeval	stamp;			/* time in past */
    double		stamp_f;		/* stamp as real number */
    int			pred = PRED_NONE;
    int			sign = 0;		/* pander to gcc on NetBSD */
    int			tmp;
    int			verbose = 0;
    char		*p;
    char		*q;
    char		path[MAXPATHLEN+1];

    /* can't use getopt() because <timespec> may begin with '-' */
    if (argc > 1 && strcmp(argv[1], "-v") == 0) {
	verbose++;
	argc--;
	argv++;
    }

    if (argc < 2) {
	usage();
	/*NOTREACHED*/
    }

    /* Start with current time, then timespec refers to an offset into
     * the past.
     */
    pmtimevalNow(&now);
    now_f = pmtimevalToReal(&now);
    stamp = now;		/* struct assignment */

    if (strcmp(argv[1], "ctime") == 0)
	pred = PRED_CTIME;
    else if (strcmp(argv[1], "mtime") == 0)
	pred = PRED_MTIME;

    switch (pred) {
	case PRED_CTIME:
	case PRED_MTIME:
	    if (argc < 3) {
		usage();
		/*NOTREACHED*/
	    }
	    p = argv[2];
	    if (*p == '+')
		sign = PLUS_TIME;
	    else if (*p == '-')
		sign = MINUS_TIME;
	    else {
		fprintf(stderr, "find-filter: Error: expect '+' or '-' to start timespec, not '%c' in \"%s\"\n", *p, argv[2]);
		exit(1);
	    }
	    p++;
	    if (*p == '\0') {
		fprintf(stderr, "find-filter: Error: missing number of days in timespec \"%s\"\n", argv[2]);
		exit(1);
	    }
	    /* start with days */
	    stamp.tv_sec -= strtol(p, &q, 10)*24*3600;
	    if (p == q) {
		fprintf(stderr, "find-filter: Error: expect number of days not '%c' after %c in timespec\"%s\"\n", *p, p[-1], argv[2]);
		exit(1);
	    }
	    if (*q == '\0') {
		/* just [+|-]D */
		break;
	    }
	    p = q;
	    if (*p != ':') {
		fprintf(stderr, "find-filter: Error: expect ':' not '%c' after days in timespec \"%s\"\n", *p, argv[2]);
		exit(1);
	    }
	    /* now expect hours */
	    p++;
	    if (*p == '\0') {
		fprintf(stderr, "find-filter: Error: missing number of hours in timespec \"%s\"\n", argv[2]);
		exit(1);
	    }
	    tmp = strtol(p, &q, 10);
	    if (p == q) {
		fprintf(stderr, "find-filter: Error: expect number of hours not '%c' in timespec \"%s\"\n", *p, argv[2]);
		exit(1);
	    }
	    if (tmp < 0 || tmp > 23) {
		fprintf(stderr, "find-filter: Error: number of hours '%d' is invalid in timespec \"%s\"\n", tmp, argv[2]);
		exit(1);
	    }
	    stamp.tv_sec -= tmp*3600;
	    if (*q == '\0') {
		/* just [+|-]D:H */
		break;
	    }
	    p = q;
	    if (*p != ':') {
		fprintf(stderr, "find-filter: Error: expect ':' not '%c' after days and hours in timespec \"%s\"\n", *p, argv[2]);
		exit(1);
	    }
	    /* now expect minutes */
	    p++;
	    if (*p == '\0') {
		fprintf(stderr, "find-filter: Error: missing number of minutes in timespec \"%s\"\n", argv[2]);
		exit(1);
	    }
	    tmp = strtol(p, &q, 10);
	    if (p == q) {
		fprintf(stderr, "find-filter: Error: expect number of minutes not '%c' in timespec \"%s\"\n", *p, argv[2]);
		exit(1);
	    }
	    if (tmp < 0 || tmp > 59) {
		fprintf(stderr, "find-filter: Error: number of minutes '%d' is invalid in timespec \"%s\"\n", tmp, argv[2]);
		exit(1);
	    }
	    if (*q != '\0') {
		fprintf(stderr, "find-filter: Error: extra text after days, hours and minutes in timespec \"%s\"\n", argv[2]);
		exit(1);
	    }
	    stamp.tv_sec -= tmp*60;
	    break;

	default:
	    usage();
	    /*NOTREACHED*/
	   
    }
    stamp_f = pmtimevalToReal(&stamp);

    /*
     * Now for each pathname on stdin, stat the file and output
     * the pathname if the file's attributes pass the predicate.
     */
    while (fgets(path, MAXPATHLEN-1, stdin) != NULL) {
	struct stat	sbuf;
	int		sts;
	struct timeval	check;		/* file's time value */
	double		check_f;
	
	/* strip trailing \n */
	for (p = path; *p; p++) {
	    if (*p == '\n') {
		*p = '\0';
		break;
	    }
	}
	sts = stat(path, &sbuf);

	if (sts < 0) {
	    if (verbose)
		fprintf(stderr, "%s: stat() failed: %s\n", path, pmErrStr(-oserror()));
	    continue;
	}

	switch (pred) {
	    case PRED_CTIME:
#if defined(HAVE_ST_MTIME_WITH_E) && defined(HAVE_STAT_TIME_T)
		    check.tv_sec = sbuf.st_ctime;
		    check.tv_usec = 0;
#elif defined(HAVE_ST_MTIME_WITH_SPEC)
		    check.tv_sec = sbuf.st_ctimespec.tv_sec;
		    check.tv_usec = sbuf.st_ctimespec.tv_nsec / 1000;
#elif defined(HAVE_STAT_TIMESTRUC) || defined(HAVE_STAT_TIMESPEC) || defined(HAVE_STAT_TIMESPEC_T)
		    check.tv_sec = sbuf.st_ctim.tv_sec;
		    check.tv_usec = sbuf.st_ctim.tv_nsec / 1000;
#else
!bozo!
#endif
		    break;

	    case PRED_MTIME:
#if defined(HAVE_ST_MTIME_WITH_E) && defined(HAVE_STAT_TIME_T)
		    check.tv_sec = sbuf.st_mtime;
		    check.tv_usec = 0;
#elif defined(HAVE_ST_MTIME_WITH_SPEC)
		    check.tv_sec = sbuf.st_mtimespec.tv_sec;
		    check.tv_usec = sbuf.st_mtimespec.tv_nsec / 1000;
#elif defined(HAVE_STAT_TIMESTRUC) || defined(HAVE_STAT_TIMESPEC) || defined(HAVE_STAT_TIMESPEC_T)
		    check.tv_sec = sbuf.st_mtim.tv_sec;
		    check.tv_usec = sbuf.st_mtim.tv_nsec / 1000;
#else
!bozo!
#endif
		    break;

	    }

	check_f = pmtimevalToReal(&check);
	if (verbose)
	    fprintf(stderr, "%s: me: %.6f (now-%.6f) timespec: %.6f\n", path, check_f, now_f-check_f, stamp_f);
	if ((sign == PLUS_TIME && check_f < stamp_f) ||
	    (sign == MINUS_TIME && check_f >= stamp_f))
	    printf("%s\n", path);
    }

    return 0;
}
